tests/test_compile_kwargs.py

```diff
@@ -0,0 +1,555 @@
+    """
+
+    X, y = make_classification()
+
+    optimizer = "sgf"  # sgf is not a loss
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer=optimizer,
+        loss="binary_crossentropy",
+    )
+    with pytest.raises(ValueError, match="Unknown optimizer"):
+        est.fit(X, y)
+
+
+def test_compiling_of_routed_parameters():
+    """Tests that routed parameters
+    can themselves be compiled.
+    """
+
+    X, y = make_classification()
+
+    class Foo:
+        got = dict()
+
+        def __init__(self, foo_kwarg):
+            self.foo_kwarg = foo_kwarg
+
+    class MyLoss:
+        def __init__(self, param1):
+            self.param1 = param1
+            self.__name__ = str(id(self))
+
+        def __call__(self, y_true, y_pred):
+            return losses_module.binary_crossentropy(y_true, y_pred)
+
+    est = KerasClassifier(
+        model=get_model,
+        loss=MyLoss,
+        loss__param1=[Foo, Foo],
+        loss__param1__foo_kwarg=1,
+        loss__param1__0__foo_kwarg=2,
+    )
+    est.fit(X, y)
+    assert est.model_.loss.param1[0].foo_kwarg == 2
+    assert est.model_.loss.param1[1].foo_kwarg == 1
+
+
+@pytest.mark.parametrize(
+    "loss", (losses_module.BinaryCrossentropy, "BinaryCrossentropy",),
+)
+@pytest.mark.parametrize("n_outputs_", (1, 2))
+def test_loss(loss, n_outputs_):
+    """Tests compiling of single loss
+    using routed parameters.
+    """
+
+    X, y = make_classification()
+    y = np.column_stack([y for _ in range(n_outputs_)]).squeeze()
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer="sgd",
+        loss=loss,
+        loss__name="custom_name",
+    )
+    est.fit(X, y)
+    assert str(loss) in str(est.model_.loss) or isinstance(
+        est.model_.loss, loss
+    )
+
+
+def test_loss_invalid_string():
+    """Tests that a ValueError is raised when an unknown
+    string is passed as a loss.
+    """
+
+    X, y = make_classification()
+
+    loss = "binary_crossentropr"  # binary_crossentropr is not a loss
+
+    est = KerasClassifier(
+        model=get_model, num_hidden=20, optimizer="sgd", loss=loss,
+    )
+    with pytest.raises(ValueError, match="Unknown loss function"):
+        est.fit(X, y)
+
+
+def test_loss_uncompilable():
+    """Tests that a TypeError is raised when a loss
+    that is not compilable is passed routed parameters.
+    """
+
+    X, y = make_classification()
+
+    loss = losses_module.binary_crossentropy
+
+    est = KerasClassifier(
+        model=get_model,
+        num_hidden=20,
+        optimizer="sgd",
+        loss=loss,
+        loss__from_logits=True,
+    )
+    expected_param = {"from_logits": True}
+    with pytest.raises(
+        TypeError,
+        match=f'TypeError: "{str(loss)}" object of type "{type(loss)}"'
+        "does not accept parameters because it's not a class."
+        f' However, it received parameters "{expected_param}"',
```

Comment:
I'd only keep the core part of the error message, `"does not accept parameters"`

Reply:
Thanks. You pointed it out earlier and I did not edit all of them.